### PR TITLE
1273 improve proving interface

### LINKED ISSUE #1273: Improve proving interface
When proving,
* **first** create "empty" ProofAttempts with the corresponding state
* when Hets returns, parse the result and fill those ProofAttempts and update the status

Also
* create an index page for ProofAttempts for Ontology[Version]

Add a ProofOptions model with
* prover
* selected axioms
* list of goals
* timeout, possibly a cumulated timeout for all the goals
* translation

### PR DESCRIPTION
**Pull request by eugenk**
This shall fix #1273, except for the index page. The index page will part of #1283.

This branch is based on the branch of #1263 (**1237-let_the_user_select_a_prover**) instead of **staging**.  It might be easier to review that one first.

### CODE REVIEW

**Comment by eugenk** [comment]
This branch is now based on staging.

**Comment by tillmo** [comment]
Coming from #1237, I tried to migrate, but got an error:
```
rake db:migrate
==  AddStateToProofAttemptsAndTheorems: migrating =============================
-- add_column(:proof_attempts, :state, :string, {:default=>"pending", :null=>false})
   -> 0.0240s
-- add_column(:proof_attempts, :state_updated_at, :datetime)
   -> 0.0006s
-- add_column(:proof_attempts, :last_error, :text)
   -> 0.0005s
-- add_column(:sentences, :state, :string, {:null=>true})
   -> 0.0008s
-- add_column(:sentences, :state_updated_at, :datetime)
   -> 0.0004s
-- add_column(:sentences, :last_error, :text)
   -> 0.0012s
==  AddStateToProofAttemptsAndTheorems: migrated (0.0279s) ====================

==  CreateProofAttemptConfigurations: migrating ===============================
-- create_table(:proof_attempt_configurations)
   -> 0.0092s
-- add_column(:proof_attempts, :proof_attempt_configuration_id, :integer, {:null=>false})
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::NotNullViolation: ERROR:  column "proof_attempt_configuration_id" contains null values
: ALTER TABLE "proof_attempts" ADD COLUMN "proof_attempt_configuration_id" integer NOT NULL/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:650:in `async_exec'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:650:in `block in execute'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract_adapter.rb:280:in `block in log'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activesupport-3.2.21/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract_adapter.rb:275:in `log'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:649:in `execute'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:1022:in `add_column'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:466:in `block in method_missing'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:438:in `block in say_with_time'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:438:in `say_with_time'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:458:in `method_missing'
/home/till/web2.0/ontohub/db/migrate/20150320135159_create_proof_attempt_configurations.rb:11:in `change'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:407:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:407:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:129:in `with_connection'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:389:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:528:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:720:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `call'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `block in ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/database_statements.rb:192:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/transactions.rb:208:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:719:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:570:in `up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:551:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/railties/databases.rake:193:in `block (2 levels) in <top (required)>'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `eval'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `<main>'
ActiveRecord::StatementInvalid: PG::NotNullViolation: ERROR:  column "proof_attempt_configuration_id" contains null values
: ALTER TABLE "proof_attempts" ADD COLUMN "proof_attempt_configuration_id" integer NOT NULL
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:650:in `async_exec'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:650:in `block in execute'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract_adapter.rb:280:in `block in log'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activesupport-3.2.21/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract_adapter.rb:275:in `log'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:649:in `execute'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:1022:in `add_column'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:466:in `block in method_missing'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:438:in `block in say_with_time'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:438:in `say_with_time'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:458:in `method_missing'
/home/till/web2.0/ontohub/db/migrate/20150320135159_create_proof_attempt_configurations.rb:11:in `change'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:407:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:407:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:129:in `with_connection'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:389:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:528:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:720:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `call'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `block in ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/database_statements.rb:192:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/transactions.rb:208:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:719:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:570:in `up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:551:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/railties/databases.rake:193:in `block (2 levels) in <top (required)>'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `eval'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `<main>'
PG::NotNullViolation: ERROR:  column "proof_attempt_configuration_id" contains null values
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:650:in `async_exec'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:650:in `block in execute'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract_adapter.rb:280:in `block in log'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activesupport-3.2.21/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract_adapter.rb:275:in `log'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:649:in `execute'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/postgresql_adapter.rb:1022:in `add_column'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:466:in `block in method_missing'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:438:in `block in say_with_time'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:438:in `say_with_time'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:458:in `method_missing'
/home/till/web2.0/ontohub/db/migrate/20150320135159_create_proof_attempt_configurations.rb:11:in `change'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:407:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:407:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:129:in `with_connection'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:389:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:528:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:720:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `call'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `block in ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/database_statements.rb:192:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/transactions.rb:208:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:719:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:570:in `up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:551:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/railties/databases.rake:193:in `block (2 levels) in <top (required)>'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `eval'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

**Comment by eugenk** [comment]
This should be fixed with 0aff746

**Comment by tillmo** [comment]
When I try to prove something, first a message "there are no proof attempts" appears. Shouldn't this be avoided, because now always a proof attempt exists immediately? See http://develop.ontohub.org/sandbox/Vickrey_proof//sp///theorems for theorems that cannot be proved within 10 secs.

**Comment by eugenk** [comment]
I think I haven't merged this branch into develop yet. Have tried this on your local ontohub instance or on develop.ontohub.org?

**Comment by tillmo** [comment]
oh sorry, it was on develop.ontohub.org. Locally, it works as expected.

File: db/migrate/20150320113702_add_state_to_proof_attempts_and_theorems.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  def change
+    add_column :proof_attempts, :state, :string, default: 'pending', null: false
+    add_column :proof_attempts, :state_updated_at, :datetime
+    add_column :proof_attempts, :last_error, :text
+
+    add_column :sentences, :state, :string, null: true # null for axioms
+    add_column :sentences, :state_updated_at, :datetime
+    add_column :sentences, :last_error, :text
+  end
+end
```
Comment (tillmo) [inline db/migrate/20150320113702_add_state_to_proof_attempts_and_theorems.rb]: I wonder why you explicitly specify `null: true` sometimes, and sometimes the same is achieved because it is the default.

Comment (eugenk) [inline db/migrate/20150320113702_add_state_to_proof_attempts_and_theorems.rb]: When I explicitly specify `null: true` it's an unusual place for a null value and I want to emphasize that null is allowed. In this case it is only supposed to be null for axioms.

File: lib/hets/prove/evaluator.rb [hunk 2]
```diff
@@ -13,17 +13,19 @@ class Evaluator
...
         self.io = io
         initialize_handling
       end
 
       def import
-        callback = ProveEvaluator.new(self)
+        callback = ProveEvaluator.new(self, proof_attempts)
         ActiveRecord::Base.transaction requires_new: true do
           callback.process(:all, :start)
           parser.parse(callback: callback)
```
Comment (tillmo) [inline lib/hets/prove/evaluator.rb]: It is hard for me to get the overall structure here. Why is there lib/hets/prove/evaluator.rb and lib/hets/prove/prove_evaluator.rb? Why does ProveEvaluator inherit from BaseEvaluator, while Evaluator does inherit anything (and seems to be unrelated to the other two)?

File: lib/hets/prove_options.rb [hunk 3]
```diff
@@ -37,6 +37,8 @@ def prepare_sentences(field)
         @options[field].map! do |sentence|
           if sentence.is_a?(Sentence)
             sentence.name
+          else
+            sentence
```
Comment (tillmo) [inline lib/hets/prove_options.rb]: this is a bad variable name, because if the object is not a `Sentence`, it is returned as `sentence`.

**Comment by tillmo** [comment]
It seems that currently, the user cannot specify a timeout. If this is correct, could you open an issue with that?

File: lib/hets/prove/evaluator.rb (hunk 2 above)
Comment (eugenk) [inline lib/hets/prove/evaluator.rb]: I created #1295 for this and I want to finish that issue this week. The explanation of the classes for now:

* `Hets::Prove::Evaluator` is the entry point for evaluating the hets output. It calls the parser and passes the callbacks to the parser.
* `Hets::Prove::ProveEvaluator < Hets::BaseEvaluator` contains those callbacks and actually fills the `ProofAttempt` objects. It inherits from `BaseEvaluator` because the `BaseEvaluator` defines a convenient callback registration mechanism.

File: lib/hets/prove_options.rb (hunk 3 above)
Comment (eugenk) [inline lib/hets/prove_options.rb]: Shall be fixed in 42ca317.

**Comment by eugenk** [comment]
Right, I created #1296.

**Comment by tillmo** [comment]
:+1:
